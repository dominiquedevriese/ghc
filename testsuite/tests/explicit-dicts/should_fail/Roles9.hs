{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE RoleAnnotations #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
module Roles9 where

import GHC.Base (Ord(..))

reverseOrd :: Ord a => Ord.Dict a
reverseOrd = undefined

data Set a
type role Set nominal

insert :: Ord a => a -> Set a -> Set a
insert = undefined

class a ~ b => Equal a b
instance Equal a a

evilInsert :: (Ord a, Equal a b) => a -> b -> Set b -> Set b
evilInsert _ = insert


f9 = evilInsert ((reverseOrd @Int :: Ord a))
