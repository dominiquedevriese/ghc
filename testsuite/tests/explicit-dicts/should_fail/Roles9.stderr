
Roles9.hs:26:6: error:
    • Explicit dictionary application to: ‘Ord a’
      is not allowed because the instance incoherence check
      requires that its type arguments do not occur in an equality,
      found: ‘a ~ b’
    • In the expression: evilInsert ((reverseOrd @Int :: Ord a))
      In an equation for ‘f9’:
          f9 = evilInsert ((reverseOrd @Int :: Ord a))
